README
/*
 *  mpc123 - Musepack Console audio player
 *  Copyright (C) 2005, 2006 Fernando Vezzosi <fvezzosi at masobit.net>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software Foundation,
 *  Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */


* Installation:
!!! Compilation requires libmpcdec-dev and libao-dev;
!!! Normal use requires libmpcdec and libao;

You can find those packages in your apt repository:
  # apt-get install libmpcdec{3,-dev} libao{,-dev}
Or on their respective websites:
  % links http://www.musepack.net/
  % links http://xiph.org/ao/

  * Downloading the bits:
    * Latest Development version (possibly heavily broken)
    Visit http://mpc123.sf.net/ for a detailed description

    * Latest Stable version
    Visit http://sf.net/projects/mpc123 to see released files, then click the
    download link

  * Compiling the bits:
    * Under GNU/Linux
    type ``make''

    * Under other OSs:
      * type ``make''
      * if any error occurs, please try to solve it and submit a patch for the
        Makefile or the sources

At this point you should have a ``mpc123'' binary. Place it where you want it
to live.

* Use of the player:
mpc123 can play files from command line arguments and from playlists.
A brief usage message is shown if you use the -h flag.

There are currently two methods to play a given mpc file:

  * Specify it on the command line
  Just type
    mpc123 /path/to/file /example/stream.mpc
  mpc123 will play the files in the specified order

  * Include it into a playlist file
  Just type
    mpc123 -@ /path/to/playlist
  mpc123 will play files specified in the file /path/to/playlist. The format
  is trivial (each line identifies a stream), you can easily generate a
  playlist with the find(1) command:

    find /path/to/musicroot -iname \*.mpc -fprint /path/to/playlist

During playback, you can have the player pass to the next playlist element (or
a random one if the -z or -Z options were specified), by sending SIGINT to the
process (this usually happens with ^C).

* Useful command line options
Each -v option increases the program verbosity. The default verbosity is 0,
which means mpc123 will produce output only in case of errors;

The -g option modifies the output volume. Allowed values range from 0 (silent)
to 100 (default value, sounds at full power); If you for example want to play
at half the sample's original gain, just use "-g 50";

The -z option randomly sorts the playlist once in the program lifetime (it is
called ``shuffle'');

The -Z option selects a random playlist element each time (this feature
instead is called ``random'');

The -q option resets the verbosity level to its default 0;

The -@ option requires an argument, specifying the path to the playlist;

* AO (Audio Output) command line options
These options control the behavior of the audio output layer;

The -o option selects the desired audio output driver (defaults to "oss").
Examples:
  mpc123 -o oss -a /dev/dsp1    # to select the second OSS sound card
  mpc123 -o alsa09              # use alsa09 as output driver, sets the output
                                  device to "default"
  mpc123 -o esd                 # use the esd daemon for output (the daemon
                                  must be running, of course ..)
  mpc123 -o null                # for debugging purposes, outputs nothing

The -a option specifies the output device for the selected driver (defaults to
"/dev/dsp")

Of course -o and -a options are related. You _can_ use only the -o option to
select the alsa09 (for example) driver, and the program will default the
output device to some usable value ("/dev/dsp" for oss, "default" for alsa)

* Hacking with mpc123
Before you start hacking with mpc123, be sure you can use tla or bazaar;
a good starting point is the arch wiki at http://wiki.gnuarch.org/

When hacking (programming and testing mpc123), be sure to turn the debug flags
on. This can be done in a couple of ways:

 1 - % make DEBUG=1
     .. compiler output ..
     %
 2 - % export DEBUG=1
     % make
     .. compiler output ..
     %

The binary file ``mpc123'' will be compiled with debugging symbols enabled,
and some useful debugging macros defined; the macros are:

- debug()
  Is to be used to produce output with debugging activated. Takes only one
  argument, which is the string to print.
- debugf()
  The same as debug(), only it takes a format string as its first argument.
  You can use it just like printf().

Both debugging macros automatically add a newline at the end of the given
string, so you don't need to add "\n".

# vi:tw=78:


ChangeLog
# do not edit -- automatically generated by arch changelog
# arch-tag: automatic-ChangeLog--bucciarati@users.sourceforge.net--2005--mpc123/mpc123--main--0.1
#

2006-08-02 11:26:35 GMT	Fernando Vezzosi <fvezzosi@masobit.net>	patch-33

    Summary:
      minor tweaks for release 0.2
    Revision:
      mpc123--main--0.1--patch-33

     * added info about the mailing list
     * cleaned Makefile out of useless stuff
     * updated TODO to reflect the actual TODO :)
    

    modified files:
     AUTHORS ChangeLog Makefile TODO


2006-08-02 10:20:20 GMT	Daniele Sempione <scrows@oziosi.org>	patch-32

    Summary:
      major/minor fix
    Revision:
      mpc123--main--0.1--patch-32

    * changed major/minor for the release :)
    

    modified files:
     ChangeLog Makefile mpc123.h


2006-07-31 18:32:42 GMT	Fernando Vezzosi <fvezzosi@masobit.net>	patch-31

    Summary:
      exit if two sigints are sent in a short time window
    Revision:
      mpc123--main--0.1--patch-31

     * now it exits if two ^C's are pressed in a rapid succession
     * fixed some bugs in the implementation ;)
    

    modified files:
     ChangeLog mpc123.c player.c playlist.c signals.c


2006-07-31 16:59:58 GMT	Daniele Sempione <scrows@oziosi.org>	patch-30

    Summary:
      sighandler is thread-safe
    Revision:
      mpc123--main--0.1--patch-30

    * now SIGINT is thread-safe
    * added support for terminating program via double SIGINT.
      now set to 0.7 secs, let's test it .. (still needed buccia's cleanup
      code, right now it only sets the flag)
    

    modified files:
     ChangeLog signals.c


2006-07-31 15:42:24 GMT	Fernando Vezzosi <fvezzosi@masobit.net>	patch-29

    Summary:
      Updated documentation (README and manpage)
    Revision:
      mpc123--main--0.1--patch-29

     * added info about esd
     * updated info about program behavior (SIGINT handling
       and choosing of default ao device)
    

    modified files:
     ChangeLog README mpc123.1


2006-07-31 15:11:09 GMT	Daniele Sempione <scrows@oziosi.org>	patch-28

    Summary:
      added 2006 to some copyrights
    Revision:
      mpc123--main--0.1--patch-28

    * ...
    

    modified files:
     ChangeLog Makefile README


2006-07-31 14:40:01 GMT	Daniele Sempione <scrows@oziosi.org>	patch-27

    Summary:
      SIGINT = next song
    Revision:
      mpc123--main--0.1--patch-27

    * now this feature works!! :))
    * ported signal handling to sigaction
    

    new files:
     .arch-ids/signals.c.id signals.c

    modified files:
     ChangeLog mpc123.c mpc123.h


2006-07-31 13:54:37 GMT	Fernando Vezzosi <fvezzosi@masobit.net>	patch-26

    Summary:
      reverted my stupid edits
    Revision:
      mpc123--main--0.1--patch-26

     * i will not mess with a thing i shouldn't have been messing with again
     * i will not use wrong string format specifiers again
    
     sorry world
    

    modified files:
     ChangeLog mpc123.h reader_file.c


2006-07-31 13:47:38 GMT	Fernando Vezzosi <fvezzosi@masobit.net>	patch-25

    Summary:
      fixed "next-song" bug, added vim modelines
    Revision:
      mpc123--main--0.1--patch-25

     * fixed a dumb bug of my previous commit, which made the player
       skip all playlist elements, when it should have skipped only
       the currently played one..
     * added the modeline "ft=c:tw=78:ts=2:et:cin" to all source files
    

    modified files:
     ChangeLog ao.c mpc123.c mpc123.h player.c playlist.c reader.c
     reader_file.c shuffle.c


2006-07-31 13:13:19 GMT	Fernando Vezzosi <fvezzosi@masobit.net>	patch-24

    Summary:
      Added macro functions for flag handling
    Revision:
      mpc123--main--0.1--patch-24

     * added flags global variable
     * added macros for flag handling
     * player loop stops if correct flags are set
    

    modified files:
     ChangeLog mpc123.h player.c


2006-07-30 10:17:38 GMT	Daniele Sempione <scrows@oziosi.org>	patch-23

    Summary:
      Makefile comment
    Revision:
      mpc123--main--0.1--patch-23

    * disables the -lefence linking option I forgot in the Makefile
    

    modified files:
     ChangeLog Makefile


2006-07-30 02:19:04 GMT	Fernando Vezzosi <fvezzosi@masobit.net>	patch-22

    Summary:
      Fixes in view of new release
    Revision:
      mpc123--main--0.1--patch-22

     * removed generation of locales in Makefile
     * better support for all sound systems, including working support for:
       - alsa09
       - esd
     * decent libao defaults are now set at runtime
     * improved man page with content and appearance
     * improved debugging output
     * cosmetic changes
    

    modified files:
     ChangeLog Makefile ao.c mpc123.1 mpc123.c mpc123.h


2006-07-29 13:38:47 GMT	Daniele Sempione <scrows@oziosi.org>	patch-21

    Summary:
      removed bugged l10n support
    Revision:
      mpc123--main--0.1--patch-21

    * l10n was too buggy, we'll re-add in the future
    * updated copyrights' year
    

    modified files:
     ChangeLog Makefile ao.c mpc123.c mpc123.h player.c playlist.c
     reader.c reader_file.c shuffle.c


2006-07-28 13:33:57 GMT	Fernando Vezzosi <fvezzosi@masobit.net>	patch-20

    Summary:
      fixed segfault when using inappropriate commandline flags
    Revision:
      mpc123--main--0.1--patch-20

     * a crappy bug that lead to sigsegv upon invocation
    

    modified files:
     ChangeLog mpc123.c


2006-07-28 13:28:17 GMT	Fernando Vezzosi <fvezzosi@masobit.net>	patch-19

    Summary:
      
    Revision:
      mpc123--main--0.1--patch-19

    
    

    modified files:
     ChangeLog Makefile mpc123.h


2006-06-06 18:39:42 GMT	Daniele Sempione <scrows@oziosi.org>	patch-18

    Summary:
      added '\n' to some strings
    Revision:
      mpc123--main--0.1--patch-18

    * ...
    

    modified files:
     ChangeLog ao.c mpc123.c mpc123.h player.c reader.c


2006-03-22 18:07:36 GMT	Fernando Vezzosi <fvezzosi@masobit.net>	patch-17

    Summary:

ao.c
/*
 *  mpc123 - Musepack Console audio player
 *  Copyright (C) 2005, 2006 Fernando Vezzosi <fvezzosi at masobit.net>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software Foundation,
 *  Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include "mpc123.h"

#include <ao/ao.h>

typedef struct mpc123_ao_data {
  ao_device * ao_dev;
  void * output_buffer;
  unsigned int output_buffer_size;
} mpc123_ao_data;

/* free a linked-list of ao_option's */
static void walk_free(ao_option * elem){
  if(elem->next){
    walk_free(elem->next);
  }
  free(elem);
}

void mpc123_choose_default_dev_by_driver(){
  /* device is already specified? then we don't need work here */
  if ( options.ao_dev ){
    debugf("WARNING: device already present (\"%s\"), but they want me"
           " to find another one .. why ??", options.ao_dev);
    return;
  }

  /* default device for oss is /dev/dsp */
  if ( !strcmp(options.ao_driver, "oss") ){
    options.ao_dev="/dev/dsp";
    return;
  }

  /* this matches "alsa" as well as "alsa09", use default card */
  if ( !strncmp(options.ao_driver, "alsa", strlen("alsa")) ){
    options.ao_dev="default";
    return;
  }

  /* other drivers don't need special device specs */
}

int mpc123_ao_init(void ** d, mpc_streaminfo * streaminfo){
  int ao_drvnum=-1;
  ao_sample_format ao_fmt;
  ao_info * ao_dinfo=NULL;
  ao_option * opt_head=NULL;
  int i;

  *d=malloc(sizeof(mpc123_ao_data));
  mpc123_ao_data * data=(mpc123_ao_data *) *d;
  
  ao_initialize();

  ao_drvnum=ao_driver_id(options.ao_driver);
  if( ao_drvnum < 0 )
    die("No suitable output driver\n");

  /* set esd options */
  if( !strcmp(options.ao_driver, "esd") ){
    opt_head=malloc(sizeof(ao_option));
    opt_head->key="esd";
    opt_head->value=options.ao_dev;
    opt_head->next=NULL;
    debugf("[ESD] ao_opts{%s}=%s", opt_head->key, opt_head->value);
  }

  /* set oss options */
  if( !strcmp(options.ao_driver, "oss") ){
    opt_head=malloc(sizeof(ao_option));
    opt_head->key="dsp";
    opt_head->value=options.ao_dev;
    opt_head->next=NULL;
    debugf("[OSS] ao_opts{%s}=%s", opt_head->key, opt_head->value);
  }

  /* set alsa options */
  if( !strcmp(options.ao_driver, "alsa09") ){
    opt_head=malloc(sizeof(ao_option));
    opt_head->key="dev";
    opt_head->value=options.ao_dev;
    opt_head->next=NULL;
    debugf("[ALSA] ao_opts{%s}=%s", opt_head->key, opt_head->value);
  }

  ao_dinfo=ao_driver_info(ao_drvnum);

  debugf("<aodriver> type=%s",
        ao_dinfo->type == AO_TYPE_LIVE ? "live" : 
        ao_dinfo->type == AO_TYPE_FILE ? "file" :
        "unknown");
  debugf("<aodriver> name=%s", ao_dinfo->name);
  debugf("<aodriver> sname=%s", ao_dinfo->short_name);
  debugf("<aodriver> comment=%s", ao_dinfo->comment);
  debug("<aodriver> options={");
  for(i=0; i<ao_dinfo->option_count; i++){
    debugf("  %s", ao_dinfo->options[i]);
  }
  debug("}");

  sayf(2, "Using ao driver \"%s\" (libao id %d)\n", ao_dinfo->name, ao_drvnum);

  /* initialize ao_format struct */
  /* XXX VERY WRONG */
  ao_fmt.bits=16;       /*tmp_stream_info.average_bitrate;*/
  ao_fmt.rate=streaminfo->sample_freq;
  ao_fmt.channels=streaminfo->channels;
  ao_fmt.byte_format=AO_FMT_LITTLE;

  /* output audio params */
  debugf("<ao_fmt> bits = %d", ao_fmt.bits);
  debugf("<ao_fmt> rate = %d", ao_fmt.rate);
  debugf("<ao_fmt> channels = %d", ao_fmt.channels);
  debugf("<ao_fmt> byte_format = %d", ao_fmt.byte_format);

#if 0
  if(strcmp(options.ao_driver,"wav")         /* ao_driver IS NOT wav */
     && strcmp(options.ao_driver,"raw")      /*              NOR raw */
     && strcmp(options.ao_driver,"au")){     /*              NOR au  */
#endif
  if(ao_dinfo->type == AO_TYPE_LIVE){
    /* live playing! */
    data->ao_dev=ao_open_live(ao_drvnum, &ao_fmt, opt_head);
  }else{
    /* output to file (act as a decoder) */
    data->ao_dev=ao_open_file(ao_drvnum, options.foutput, TRUE, &ao_fmt, NULL);
  }
  
  if( !data->ao_dev ){
    dief("Could not open audio output: [%d] %s\n", errno, strerror(errno));
  }

  if(opt_head)
    walk_free(opt_head);

  return 0;
}

#ifdef MPC_FIXED_POINT
inline static int shift_signed(MPC_SAMPLE_FORMAT val, int shift)
{
  if (shift > 0)
    val <<= shift;
  else if (shift < 0)
    val >>= -shift;
  return (int) val;
}
#endif

/* play buffer[samples * 2] on device data->ao_dev */
int mpc123_ao_play(void * d, MPC_SAMPLE_FORMAT * buffer, unsigned samples){
  mpc123_ao_data * data=(mpc123_ao_data *) d;
  unsigned n; int val;
  unsigned char * output;
  unsigned bytes=samples * 2;   /* 16 bit == 2 bytes */

  const unsigned pSize = 16;
  const int clip_min    = -1 << (pSize - 1);
  const int clip_max    = (1 << (pSize - 1)) - 1;
  const int float_scale =  1 << (pSize - 1);

  /* if the buffer is already allocated, use it */
  output = data->output_buffer ? (unsigned char *) data->output_buffer :
          (data->output_buffer = (unsigned char *) malloc(bytes));

  /* realloc if the size changed */
  if(data->output_buffer_size != bytes){
    debugf("osize=%d nsize=%d", data->output_buffer_size, bytes);
    output = data->output_buffer = realloc(data->output_buffer, bytes);
    data->output_buffer_size = bytes;
  }

  if(!output){
    dief("Out of memory (needed 0x%08x bytes)\n", bytes);
  }

  /* code stolen from xmms-musepack */
  for (n = 0; n < samples; n++) {
#ifdef MPC_FIXED_POINT
    val = shift_signed(buffer[n], pSize - MPC_FIXED_POINT_SCALE_SHIFT);
#else
    /* adjust gain */
    buffer[n] *= options.volume;

    /* convert to pcm */
    val = (int) (buffer[n] * float_scale);
#endif
    if (val < clip_min)
      val = clip_min;
    else if (val > clip_max)
      val = clip_max;

    /* store pcm bytes in buffer, in the right order */
    output[n * 2] = (unsigned char) (val & 0xFF);
    output[n * 2 + 1] = (unsigned char) ((val >> 8) & 0xFF);
  }

  /* ao_play(ao_device *, void *, uint_32); */
  val=ao_play(data->ao_dev, (void *)output, bytes);
  return val;
}

/* called when done playing */
int mpc123_ao_done(void * d){
  mpc123_ao_data * data=(mpc123_ao_data *) d;

  ao_close(data->ao_dev);
  free(data);

  return 0;
}

/* vim:ft=c:tw=78:ts=2:et:cin:
 */
